x/concentrated-liquidity/lp_test.go

```diff
@@ -2,66 +2,252 @@ package concentrated_liquidity_test
+	expectedError   error
+}
 
-	tests := map[string]struct {
-		poolId            uint64
-		currentTick       sdk.Int
-		lowerTick         int64
-		upperTick         int64
-		currentSqrtP      sdk.Dec
-		amount0Desired    sdk.Int
-		amount0Expected   sdk.Int
-		amount1Desired    sdk.Int
-		amount1Expected   sdk.Int
-		expectedLiquidity sdk.Dec
-	}{
-		"happy path": {
-			poolId:            1,
-			currentTick:       sdk.NewInt(85176),
-			lowerTick:         int64(84222),
-			upperTick:         int64(86129),
-			currentSqrtP:      sdk.MustNewDecFromStr("70.710678118654752440"), // 5000
-			amount0Desired:    sdk.NewInt(1000000),                            // 1 eth
-			amount0Expected:   sdk.NewInt(998587),                             // 0.998587 eth
-			amount1Desired:    sdk.NewInt(5000000000),                         // 5000 usdc
-			amount1Expected:   sdk.NewInt(5000000000),                         // 5000 usdc
-			expectedLiquidity: sdk.MustNewDecFromStr("1517818840.967515822610790519"),
+var (
+	denom0   = "eth"
+	denom1   = "usdc"
+	baseCase = &lpTest{
+		poolId:          1,
+		currentTick:     sdk.NewInt(85176),
+		lowerTick:       int64(84222),
+		upperTick:       int64(86129),
+		currentSqrtP:    sdk.MustNewDecFromStr("70.710678118654752440"), // 5000
+		amount0Desired:  sdk.NewInt(1000000),                            // 1 eth
+		amount0Expected: sdk.NewInt(998587),                             // 0.998587 eth
+		amount1Desired:  sdk.NewInt(5000000000),                         // 5000 usdc
+		amount1Expected: sdk.NewInt(5000000000),                         // 5000 usdc
+		liquidityAmount: sdk.MustNewDecFromStr("1517818840.967515822610790519"),
+	}
+)
+
+func (s *KeeperTestSuite) TestCreatePosition() {
+	tests := map[string]lpTest{
+		"base case": *baseCase,
+		"error: invalid tick": {
+			lowerTick:     types.MaxTick + 1,
+			expectedError: types.InvalidTickError{Tick: types.MaxTick + 1, IsLower: true},
 		},
+		// TODO: add more tests
+		// - custom hand-picked values
+		// - error edge cases
+		// - think of overflows
+		// - think of truncations
 	}
 
 	for name, tc := range tests {
 		s.Run(name, func() {
+			tc := tc
 			s.SetupTest()
 
+			// Merge tc with baseCase and update tc
+			// to the merged result. This is done
+			// to reduce the amount of boilerplate
+			// in test cases.
+			baseConfigCopy := *baseCase
+			mergeConfigs(&baseConfigCopy, &tc)
+			tc = baseConfigCopy
+
 			s.App.ConcentratedLiquidityKeeper.CreateNewConcentratedLiquidityPool(s.Ctx, tc.poolId, denom0, denom1, tc.currentSqrtP, tc.currentTick)
 
 			asset0, asset1, liquidityCreated, err := s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, tc.poolId, s.TestAccs[0], tc.amount0Desired, tc.amount1Desired, sdk.ZeroInt(), sdk.ZeroInt(), tc.lowerTick, tc.upperTick)
+
+			if tc.expectedError != nil {
+				s.Require().Error(err)
+				s.Require().Equal(asset0, sdk.Int{})
+				s.Require().Equal(asset1, sdk.Int{})
+				s.Require().ErrorAs(err, &tc.expectedError)
```

Comment:
nit: probably can just go with ErrorAs instead of also requiring Error

Reply:
https://github.com/osmosis-labs/osmosis/pull/3459#discussion_r1026994637